articles/embedded.md

```diff
@@ -0,0 +1,309 @@
+DDS, simple (custom) serial communication protocols just as common 
+wireless technologies like 6LoWPAN
+
+### W-MODULARITY: *High Modularity* 
+
+The embedded ROS 2 stack shall be highly modular, not only to allow for 
+adaptation to specific hardware features and different RTOS, but also to
+allow the integration with existing frameworks and to be able to create 
+customized sub-stacks. For example, it shall be usable in the style of 
+rosserial, thus bringing basically the message serialization to the MCU 
+only. Similarly, it should be possible to derive a sub-stack that provides 
+node graph features and publish/subscribe but no parameters and services, 
+and so on.
+
+### W-CONTROL: *Support control-oriented applications*  
+
+MCUs are great for control, and micro-ROS should be as well. This usually 
+means hard real-time performance with low jitter and small response times.
+
+### W-POWER *Make low-power modes possible* 
+
+MCUs are often used in battery-powered applications, and/or in applications
+with a large amount of standby time. Therefore, the stack should make it 
+easily possible to save power.
+
+### W-STATIC: *Use static initialization* 
+
+Static initialization is less error-prone, easier to analyze, and shifts 
+memory usage from RAM to flash. It is a requirement for going to very small 
+resource use. 
+
+### W-BOOT: *Quick boot times compared to native ROS 2 machines* 
+
+Microcontrollers provide with the capability of booting very quickly
+(typically, in the order of tenths of milliseconds or less). This enhances
+the existing landscape of ROS 2-native devices which in the best cases, 
+require a few seconds to boot an embedded Linux.
+
+### W-INFOMODEL: *A common interface that facilitates interoperability across different vendors* 
+
+An information model for robot devices, adapted by vendors of robot modules, 
+would lower costs of integration of robotic systems. Within the OFERA project 
+we propose and use the 
+[Hardware Robot Information Model (HRIM)](https://acutronicrobotics.com/modularity/hrim/).|
+
+### More Details / Ideas
+
+In the EU project OFERA, we compiled a longer list of requirements to an embedded 
+ROS 2 stack, which might serve as a good basis for the discussion. Please see 
+Section 5 in 
+[OFERA deliverable D1.7_Requirements.pdf](http://ofera.eu/storage/deliverables/OFERA_D1.7_Requirements.pdf).
+
+
+## Questions
+
+To determine how feasible that is, and to come up with the actual requirements and/or design, a few questions come to mind. I've cross-referenced them to the relevant wishlist items (most of the there's more than one -- this suggests areas where trade-offs may be required, or alternatives pursued).
+
+| Key | Related Wish | Question |
+|-----|-------------|----------|
+
+### Q-RTOS: *Which RTOS(s) do we use as the basis?*
+
+#### Related: W-RTOS, W-PORTABILITY
+
+### Q-BUILD: *How do we handle the RTOS(s) respective build-systems?*
```

Comment:
Can plugins for colcon support them? Although I know that there are many embedded devices with annoying Windows-only tool chains.

Reply:
Not sure. The general issue is that you might need headers from the RTOS, and those can come with a lot of preprocessor magic set by their build systems. It *may* be possible to create a config.h or similar, but this remains to be investigated.

Here's what I know:
 * For FreeRTOS, you usually include it in the app, so no issue there
 * For Arduino, you could invoke Arduino's build_platfrom from colcon, but it expects some stuff to be *copied* to the Arduino workspace, so you at least need to know needs to be copied there
 * NuttX has its own build-system similar (or even the same, not sure) as the Linux one. It creates a config.h and then uses a regular Makefile infrastructure to compile. It also expects applications to reside in an "apps" directory. There's also some dynamic loading capability, but not sure how that works.